File: day18/src/main.py
from heapq import heappop, heappush
from pathlib import Path


def parse_input(filename):
    return [tuple(map(int, line.split(','))) for line in Path(filename).read_text().strip().splitlines()]


def find_shortest_path(grid, grid_size):
    start = (0, 0)
    target = (grid_size - 1, grid_size - 1)
    directions = [(0, -1), (0, 1), (-1, 0), (1, 0)]

    visited = [row[:] for row in grid]
    priority_queue = [(0, start, [start])]

    while priority_queue:
        _, (x, y), path = heappop(priority_queue)

        if (x, y) == target:
            return path

        if visited[y][x]:
            continue

        visited[y][x] = True

        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if 0 <= nx < grid_size and 0 <= ny < grid_size and not visited[ny][nx]:
                heappush(priority_queue, (len(path), (nx, ny), path + [(nx, ny)]))

    return []


def part1():
    grid_list = parse_input('my_input.txt')
    grid_size = 71
    byte_limit = 1024

    grid = [[False for _ in range(grid_size)] for _ in range(grid_size)]
    for x, y in grid_list[:byte_limit]:
        grid[y][x] = True

    path = find_shortest_path(grid, grid_size)

    print(f'Answer: {len(path) - 1}')


def part2():
    grid_list = parse_input('my_input.txt')
    grid_size = 71

    grid = [[False for _ in range(grid_size)] for _ in range(grid_size)]
    path = find_shortest_path(grid, grid_size)

    for i, (x, y) in enumerate(grid_list):
        print(f'Progress: {i * 100 // len(grid_list)}%', end='\r', flush=True)

        grid[y][x] = True

        # only re-run simulation if new corrupted position affects the path
        if (x, y) in path:
            path = find_shortest_path(grid, grid_size)

            if not path:
                break

    print('\r\033[2K', end='\r')

    print(f'Answer: {x},{y}')


if __name__ == '__main__':
    part1()
    part2()

File: day18/src/main.rs
use std::collections::BinaryHeap;
use std::fs;

type Point = (usize, usize);

fn parse_input(filename: &str) -> Vec<Point> {
    fs::read_to_string(filename)
        .expect("Failed to read file")
        .trim()
        .lines()
        .map(|line| {
            let parts: Vec<usize> = line
                .split(',')
                .map(|s| s.parse().unwrap())
                .collect();
            (parts[0], parts[1])
        })
        .collect()
}

fn find_shortest_path(grid: &Vec<Vec<bool>>, grid_size: usize) -> Vec<Point> {
    let start: Point = (0, 0);
    let target: Point = (grid_size - 1, grid_size - 1);
    let directions = [(0, -1), (0, 1), (-1, 0), (1, 0)];

    let mut visited = vec![vec![false; grid_size]; grid_size];
    let mut priority_queue = BinaryHeap::from(vec![(0, start, vec![start])]);

    while let Some((_, (x, y), path)) = priority_queue.pop() {
        if (x, y) == target {
            return path;
        }

        if visited[y][x] {
            continue;
        }

        visited[y][x] = true;

        for &(dx, dy) in &directions {
            let nx = x.wrapping_add(dx as usize);
            let ny = y.wrapping_add(dy as usize);
            if nx < grid_size && ny < grid_size && !visited[ny][nx] && !grid[ny][nx] {
                let mut new_path = path.clone();
                new_path.push((nx, ny));
                priority_queue.push((usize::MAX - new_path.len(), (nx, ny), new_path));
            }
        }
    }

    vec![]
}

fn part1() {
    let grid_list = parse_input("my_input.txt");
    let grid_size = 71;
    let byte_limit = 1024;

    let mut grid = vec![vec![false; grid_size]; grid_size];
    for &(x, y) in &grid_list[..byte_limit] {
        grid[y][x] = true;
    }

    let path = find_shortest_path(&grid, grid_size);

    println!("Answer: {}", path.len() - 1);
}

fn part2() {
    let grid_list = parse_input("my_input.txt");
    let grid_size = 71;

    let mut grid = vec![vec![false; grid_size]; grid_size];
    let mut path = find_shortest_path(&grid, grid_size);

    for (i, &(x, y)) in grid_list.iter().enumerate() {
        print!("\rProgress: {}%", i * 100 / grid_list.len());
        grid[y][x] = true;

        // only re-run simulation if new corrupted position affects the path
        if path.contains(&(x, y)) {
            path = find_shortest_path(&mut grid, grid_size);

            if path.is_empty() {
                print!("\r\x1b[2K");
                println!("Answer: {},{}", x, y);
                break;
            }
        }
    }
}

fn main() {
    part1();
    part2();
}
